### PR TITLE
Fix: withdraw pending balance

### PR DESCRIPTION
**Pull request by fcavazzoli**
# Description

 - Add magic_bride call in dip20_proxy to get token canister id
 - add ClaimableMessage if send message is success for both eth and dpi20 proxy

## Changes
 - New attributes definition for Withdraw `withdraw(eth_contract_as_principal: TokendId, eth_addr: EthereumAddr, _amount: Nat,)`
 - New struct `pub struct WithdrawableBalance(pub Vec<(String, String, Nat)>)`

### CODE REVIEW

File: magic_bridge/ic/src/dip20_proxy/src/api/withdraw.rs [hunk 1]
```diff
@@ -1,33 +1,52 @@
...
 };
 
 /// withdraw left over balance if burn/mint fails
 /// this will attempt to bridge the leftover balance
 /// todo withdraw specific balance
 #[update(name = "withdraw")]
 #[candid_method(update, rename = "withdraw")]
-pub async fn withdraw(token_id: TokendId, eth_addr: EthereumAddr, _amount: Nat) -> TxReceipt {
+pub async fn withdraw(
+    eth_contract_as_principal: TokendId,
```
Comment (b0xtch) [inline magic_bridge/ic/src/dip20_proxy/src/api/withdraw.rs:25]: should we expect users to pass in the eth contract instead of canister tokenId?

Comment (fcavazzoli) [reply magic_bridge/ic/src/dip20_proxy/src/api/withdraw.rs:25]: yes, same as `Burn`

**Review by b0xtch — APPROVED** [review]

File: eth_bridge/ic/src/eth_proxy/src/common/types.rs [hunk 2]
```diff
@@ -77,7 +80,7 @@ pub struct ProxyState {
     /// store incoming messages against status locks
     pub incoming_messages: RefCell<HashMap<MessageHash, MessageStatus>>,
     /// user balances
-    pub balances: RefCell<HashMap<Principal, HashMap<TokendId, Nat>>>,
+    pub balances: RefCell<HashMap<Principal, Vec<(Principal, Nat)>>>,
```
Comment (b0xtch) [inline eth_bridge/ic/src/eth_proxy/src/common/types.rs:83]:
```suggestion
    pub balances: RefCell<HashMap<EthereumAddr, Nat>>,
```
Could we narrow it down to this?

File: eth_bridge/ic/src/eth_proxy/src/common/types.rs [hunk 3]
```diff
@@ -91,7 +94,7 @@ pub struct StableProxyState {
     /// store incoming messages against status locks
     pub incoming_messages: HashMap<MessageHash, MessageStatus>,
     /// user balances
-    pub balances: HashMap<Principal, HashMap<Principal, Nat>>,
+    pub balances: Option<HashMap<Principal, Vec<(Principal, Nat)>>>,
```
Comment (b0xtch) [inline eth_bridge/ic/src/eth_proxy/src/common/types.rs:97]:
```suggestion
pub balances: RefCell<HashMap<EthereumAddr, Nat>>,
```
The same usage can be applied here

File: eth_bridge/ic/src/eth_proxy/src/common/types.rs (hunk 2 above)
Comment (b0xtch) [reply eth_bridge/ic/src/eth_proxy/src/common/types.rs:83]: If we need `caller` then ignore

File: eth_bridge/ic/src/eth_proxy/src/proxy.rs [hunk 4]
```diff
@@ -38,44 +36,58 @@ impl ProxyState {
...
-            .add_assign(amount.clone())
+    pub fn add_balance(&self, caller: Principal, to: Principal, amount: Nat) {
+        let mut binding = self.balances.borrow_mut();
+        let balance = binding.entry(caller).or_default();
+        match balance.into_iter().find(|tx| tx.0 == to) {
+            Some(value) => value.1 += amount,
+            None => {
+                balance.push((to, amount));
+            }
+        }
```
Comment (b0xtch) [inline eth_bridge/ic/src/eth_proxy/src/proxy.rs:76]: using a `hashmap` in the original structure could simplify this a lot

```rust
RefCell<HashMap<Principal, HashMap<EthereumAddr, Nat>>>
```

File: eth_bridge/ic/src/eth_proxy/src/proxy.rs [hunk 5]
```diff
@@ -38,44 +36,58 @@ impl ProxyState {
...
+        let mut binding = self.balances.borrow_mut();
+        let user_balances = binding.get_mut(&caller).unwrap();
+        // if new amount is zero, we remove the tx.
+        if amount == 0 {
+            let index = user_balances.into_iter().position(|tx| tx.0 == to).unwrap();
+            user_balances.remove(index);
+        } else {
+            let mut balance = user_balances.iter_mut().find(|tx| tx.0 == to).unwrap();
+            balance.1 = amount;
+        }
```
Comment (b0xtch) [inline eth_bridge/ic/src/eth_proxy/src/proxy.rs:90]: same as above, simplifying our structure

File: eth_bridge/ic/src/eth_proxy/src/common/types.rs (hunk 3 above)
Comment (b0xtch) [reply eth_bridge/ic/src/eth_proxy/src/common/types.rs:97]: Also Is `Optional` being committed or only required for upgrades?

File: magic_bridge/ic/src/dip20_proxy/src/common/types.rs [hunk 6]
```diff
@@ -78,7 +81,7 @@ pub struct ProxyState {
     /// store incoming messages against status locks
     pub incoming_messages: RefCell<HashMap<MessageHash, MessageStatus>>,
     /// user balances
-    pub balances: RefCell<HashMap<Principal, HashMap<TokendId, Nat>>>,
+    pub balances: RefCell<HashMap<Principal, HashMap<TokendId, Vec<(Principal, Nat)>>>>,
```
Comment (b0xtch) [inline magic_bridge/ic/src/dip20_proxy/src/common/types.rs:84]: Same idea as the `eth_proxy` balances structure.

- Simplify
- and if caller is not needed truncate to:

```rust
EthereumAddr
RefCell<HashMap<EthereumAddr, HashMap<TokenId, , Nat>>>,
```

File: magic_bridge/ic/src/dip20_proxy/src/proxy.rs [hunk 7]
```diff
@@ -38,44 +37,88 @@ impl ProxyState {
...
-    pub fn get_balance(&self, caller: Principal, token_id: TokendId) -> Option<Nat> {
-        self.balances
-            .borrow()
-            .get(&caller)
-            .map(|s| s.get(&token_id))
-            .map(|b| match b {
-                Some(balance) => balance.clone(),
-                None => Nat::from(0_u32),
-            })
+    pub fn get_balance(
```
Comment (b0xtch) [inline magic_bridge/ic/src/dip20_proxy/src/proxy.rs:40]: this will be simplified, if the structure is also simplified

**Review by b0xtch — COMMENTED** [review]
A few comments regarding data structures and one spelling mistake across the project `TokendId` -> `TokenId`

File: eth_bridge/ic/src/eth_proxy/src/common/types.rs (hunk 2 above)
Comment (fcavazzoli) [reply eth_bridge/ic/src/eth_proxy/src/common/types.rs:83]: we need the caller

File: eth_bridge/ic/src/eth_proxy/src/common/types.rs (hunk 3 above)
Comment (fcavazzoli) [reply eth_bridge/ic/src/eth_proxy/src/common/types.rs:97]: it is required for the upgrades

**Pull request merged**